torchao/prototype/mx_formats/nvfp4_tensor.py

```diff
@@ -0,0 +1,538 @@
+        end_block = None if end is None else end // x._block_size
+
+        # Slice the scale tensor accordingly
+        sliced_scale = aten.slice.Tensor(scale_shaped, 1, start_block, end_block, step)
+    else:
+        raise ValueError(
+            f"NVFP4Tensor only supports slicing along dimensions 0 and 1, got dim={dim}"
+        )
+
+    return NVFP4Tensor(
+        sliced_scale,
+        x._per_tensor_scale,  # Unchanged per-tensor scale
+        sliced_data,
+        x._block_size,
+        x._orig_dtype,
+    )
+
+
+@implements([aten.t.default])
+def nvfp4_t(func, types, args, kwargs):
+    # For now, only transpose(input, 0, 1) is supported.
+    old = args[0]
+    new = NVFP4Tensor(
+        old._scale_e4m3,
+        old._per_tensor_scale,
+        old._data.t(),
+        old._block_size,
+        old._orig_dtype,
+    )
+    return new
+
+
+@implements([aten.view.default])
+def nvfp4_view_op(func, types, args, kwargs):
+    data = args[0]._data
+    new_size = args[1]
+    new_size = tensor_size_hp_to_fp4x2(new_size, data.is_contiguous())
+    new_data = func(data, new_size, *args[2:], **kwargs)
+    return NVFP4Tensor(
+        args[0]._scale_e4m3,
+        args[0]._per_tensor_scale,
+        new_data,
+        args[0]._block_size,
+        args[0]._orig_dtype,
+    )
+
+
+def _addmm_nvfp4_dispatch(
+    a: NVFP4Tensor, b: NVFP4Tensor, aten_op, bias: Optional[torch.Tensor] = None
+) -> torch.Tensor:
+    """
+    Core implementation shared between nvfp4_mm, nvfp4_addmm, and nvfp4_linear.
+    The only difference is whether bias is None or not.
+    """
+    assert a._data.is_contiguous()
+    assert b._data.t().is_contiguous()
+    assert a._block_size == 16, f"NVFP4 requires block_size=16, got {a._block_size}"
+    assert b._block_size == 16, f"NVFP4 requires block_size=16, got {b._block_size}"
+
+    # Get tensor dimensions
+    M, K = a.shape[0], a.shape[1]
+    N = b.shape[1]
+
+    # Swizzle Dizzle
+    a_scale = a._scale_e4m3.view(M, K // a._block_size)
+    b_scale = b._scale_e4m3.view(N, K // b._block_size)
+    a_scale_blocked = to_blocked(a_scale)
+    b_scale_blocked = to_blocked(b_scale)
+
+    # Merge double quant scales into 1 scale for Scale_In^D
+    if a._per_tensor_scale is not None:
+        assert b._per_tensor_scale is not None
+        scale_result = a._per_tensor_scale * b._per_tensor_scale
+    else:
+        assert b._per_tensor_scale is None and a._per_tensor_scale is None
+        scale_result = None
+
+    result = torch._scaled_mm(
+        a._data.view(torch.float4_e2m1fn_x2),
+        b._data.view(torch.float4_e2m1fn_x2),
+        a_scale_blocked.view(torch.float8_e4m3fn),
+        b_scale_blocked.view(torch.float8_e4m3fn),
+        bias=bias,
+        out_dtype=a._orig_dtype,
+        # scale_result=scale_result,
+    )
+
+    # THIS IS A WORKAROUND:
+    # RuntimeError: CUDA error: CUBLAS_STATUS_INVALID_VALUE when calling
+    if scale_result is not None:
+        result = result * scale_result
```

Comment:
I think if there is `bias`, `scale_result` should be applied before `bias`?

Reply:
You are totally right